From 9cfe1914e00e0ce09cae2f6715462c7fff495e64 Mon Sep 17 00:00:00 2001
From: Joachim Falk <joachim.falk@gmx.de>
Date: Thu, 13 Oct 2011 19:50:32 +0200
Subject: [PATCH 3/7] Add RH patch tigervnc11-rh628054-xorg.patch which fixes vmware keyboard interaction bug.

Prior to this update, Xvnc (the X VNC server; part of the tigervnc
package) did not pass keyboard input to a remote VMware workstation
because it did not take into account types of keyboards which do not
have modifier keys. With this update, Xvnc recognizes all types of
keyboards; thus, keyboard input is correctly passed to remote VMware
workstations. (BZ#628054)
---
 unix/xserver/dix/inpututils.c |   24 +++++++++++++-----------
 1 files changed, 13 insertions(+), 11 deletions(-)

diff --git a/unix/xserver/dix/inpututils.c b/unix/xserver/dix/inpututils.c
index 4848c1b..95f903d 100644
--- a/unix/xserver/dix/inpututils.c
+++ b/unix/xserver/dix/inpututils.c
@@ -287,7 +287,7 @@ int generate_modkeymap(ClientPtr client, DeviceIntPtr dev,
 {
     CARD8 keys_per_mod[8];
     int max_keys_per_mod;
-    KeyCode *modkeymap;
+    KeyCode *modkeymap = NULL;
     int i, j, ret;
 
     ret = XaceHook(XACE_DEVICE_ACCESS, client, dev, DixGetAttrAccess);
@@ -311,18 +311,20 @@ int generate_modkeymap(ClientPtr client, DeviceIntPtr dev,
         }
     }
 
-    modkeymap = xcalloc(max_keys_per_mod * 8, sizeof(KeyCode));
-    if (!modkeymap)
-        return BadAlloc;
+    if (max_keys_per_mod != 0) {
+        modkeymap = xcalloc(max_keys_per_mod * 8, sizeof(KeyCode));
+        if (!modkeymap)
+            return BadAlloc;
 
-    for (i = 0; i < 8; i++)
-        keys_per_mod[i] = 0;
+        for (i = 0; i < 8; i++)
+            keys_per_mod[i] = 0;
 
-    for (i = 8; i < MAP_LENGTH; i++) {
-        for (j = 0; j < 8; j++) {
-            if (dev->key->xkbInfo->desc->map->modmap[i] & (1 << j)) {
-                modkeymap[(j * max_keys_per_mod) + keys_per_mod[j]] = i;
-                keys_per_mod[j]++;
+        for (i = 8; i < MAP_LENGTH; i++) {
+            for (j = 0; j < 8; j++) {
+                if (dev->key->xkbInfo->desc->map->modmap[i] & (1 << j)) {
+                    modkeymap[(j * max_keys_per_mod) + keys_per_mod[j]] = i;
+                    keys_per_mod[j]++;
+                }
             }
         }
     }
-- 
1.7.2.5

